### PR TITLE
refactor: form dialog

### PR DESCRIPTION
**Pull request by IlliaDerevianko**
- Extract dialog and form logic into separate components
- Use these components for AdminScoringForm

Screenshot 2024-09-11 at 13 27 19
*Dialog close button*

### CODE REVIEW

File: components/GenericFormDialog.tsx
Comment (procaconsul) [inline components/GenericFormDialog.tsx:1]: (1) This is overall _just_ a Dialog, if you think about it. How about:

- calling this `GenericDialog`
- turn `children` to `trigger`
- turn `form` to `children` and generalise the logic to decouple it from forms altogether

File: components/FormInDialog.tsx
Comment (procaconsul) [inline components/FormInDialog.tsx:1]: (2) Here I would: 

- collapse `close` and `closeOnSuccess` in a single prop called `onSuccess`, i.e. an _optional_ callback to call on success which defaults to `() => {}`
- rename the component to `FormWrapper`, and decouple it from the notion of dialog (as at this point there's no tie to a dialog: it's just a super Form that handles an action seamlessly and places the buttons at the bottom of the form

File: components/FormInDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        </Callout.Root>
+      )}
+      <form action={formAction} onSubmit={() => setPending(true)}>
+        {children}
+        <Flex justify="end" gap="2" mt="4">
+          <Button type="button" variant="soft" color="gray" onClick={close}>
+            Cancel
+          </Button>
+          <Button type="submit" disabled={pending}>
+            {pending ? <Spinner /> : (submitButtonText ?? 'Save')}{' '}
```
Comment (procaconsul) [inline components/FormInDialog.tsx:60]:
```suggestion
            {pending ? <Spinner /> : (submitButtonText ?? 'Save')}
```

I would also extract 'Save' in a constant, like DEFAULT_SUBMIT_BTN_TEXT

File: components/AdminScoringDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+const AdminScoringDialog: FC<AdminScoringDialogProps> = ({ row }) => {
+  const formRenderer = ({ close }: { close: () => void }) => (
+    <AdminScoringForm close={close} row={row} />
+  )
+
+  return (
+    <GenericFormDialog title="Admin Scoring Form" form={formRenderer}>
+      <Button>Admin Scoring Form</Button>
+    </GenericFormDialog>
+  )
```
Comment (procaconsul) [inline components/AdminScoringDialog.tsx:154]: (3) I would keep this component but:

- change `row` to `formData`
- extract at THIS level the form action i.e. `upsertAdminScoringWithId`
- compose the components as below:

```
<GenericDialog>
    <FormWrapper action={upsertAdminScoringWithId}>
        <AdminScoringForm data={data}/>
     </FormWrapper>
</GenericDialog>
```

**Review by procaconsul — CHANGES_REQUESTED** [review]
Nice work! I have left plenty of meaty comments. I have numbered them 1 to 3 so it's easier to follow the logic behind the suggestions.

File: components/GenericFormDialog.tsx
Comment (IlliaDerevianko) [reply components/GenericFormDialog.tsx:1]: Regarding turning 'form' to 'children', I would need to pass the close function to children. Is there a nice way to pass props to 'children' and keep its type as ReactNode?

Comment (procaconsul) [reply components/GenericFormDialog.tsx:1]: Oooops you are right, sorry. I have expaned my point (3) to answer this question.

**Review by procaconsul — CHANGES_REQUESTED** [review]
More generic comment regarding style: I'll stay consistent with direct imports from React  e.g.

- `FC` instead of `React.FC`
- `useState` instead of `React.useState` etc.

**Pull request merged**